### ISSUE TITLE
[JENKINS-1613] Ability to build on configurable subset of matrix configurations

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/matrix-project-plugin#328]
<p>It is useful sometimes to perform build on predefined number of slave builders <br/>
assigned to multi-configuration job. For example, if some OS builds failed, and <br/>
we wish to re-build only them after code fixed</p>

---
<details><summary><i>Originally reported by <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jazzjack">jazzjack</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613" target="_blank">Ability to build on configurable subset of matrix configurations</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/wolfgang.jpg" title="wolfgang's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=wolfgang">wolfgang</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>component(s)</b>: matrix-project-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2011-09-12T02:48:12+00:00
<li><b>votes</b>: 12
<li><b>watchers</b>: 11
<li><b>imported</b>: 2025-12-06
</ul></i>
<details><summary>Raw content of original issue</summary>

<pre>
<p>It is useful sometimes to perform build on predefined number of slave builders 
assigned to multi-configuration job. For example, if some OS builds failed, and 
we wish to re-build only them after code fixed</p></pre>
</details>
</details>
<ul><li><i>environment</i>: <code>Platform: All, OS: All</code></li></ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mdonohue:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=124440&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-124440">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=124440&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-124440">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This problem isn't exclusive to slave selection, as I understand the matrix
build feature, so I've updated the summary. </p>

<p>Also see issue 3332, which would allow a canary configuration to run first, and
only run the rest of the matrix configurations if the canary succeeds.  This
would cut down on massive failure notification.</p>
</pre>
</details>

</ul>
</details>

<p>This problem isn't exclusive to slave selection, as I understand the matrix<br/>
build feature, so I've updated the summary. </p>

<p>Also see issue 3332, which would allow a canary configuration to run first, and<br/>
only run the rest of the matrix configurations if the canary succeeds.  This<br/>
would cut down on massive failure notification.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mdonohue:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=124441&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-124441">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=124441&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-124441">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<ul>
	<li>
	<ul>
		<li>
		<ul>
			<li>Issue 2353 has been marked as a duplicate of this issue. ***</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

</pre>
</details>

</ul>
</details>

<ul>
	<li>
	<ul>
		<li>
		<ul>
			<li>Issue 2353 has been marked as a duplicate of this issue. ***</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mdonohue:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=124442&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-124442">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=124442&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-124442">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<ul>
	<li>
	<ul>
		<li>
		<ul>
			<li>Issue 3867 has been marked as a duplicate of this issue. ***</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

</pre>
</details>

</ul>
</details>

<ul>
	<li>
	<ul>
		<li>
		<ul>
			<li>Issue 3867 has been marked as a duplicate of this issue. ***</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=124443&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-124443">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=124443&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-124443">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in hudson
User: : huybrechts
Path:
 trunk/hudson/main/core/src/main/java/hudson/matrix/MatrixBuild.java
 trunk/hudson/main/core/src/main/java/hudson/matrix/MatrixProject.java
 trunk/hudson/main/core/src/main/java/hudson/model/Result.java
 trunk/hudson/main/core/src/main/resources/hudson/matrix/MatrixProject/configure-entries.jelly
 trunk/hudson/main/test/src/main/java/org/jvnet/hudson/test/FailureBuilder.java
 trunk/hudson/main/test/src/main/java/org/jvnet/hudson/test/UnstableBuilder.java
 trunk/hudson/main/test/src/main/resources/org/jvnet/hudson/test/UnstableBuilder/config.jelly
 trunk/hudson/main/test/src/test/java/hudson/matrix/MatrixProjectTest.java
 trunk/hudson/main/war/resources/help/matrix/touchstone.html
<a href="http://fisheye4.cenqua.com/changelog/hudson/?cs=20737" class="external-link" target="_blank" rel="nofollow noopener">http://fisheye4.cenqua.com/changelog/hudson/?cs=20737</a>
Log:
<span class="error">&#91;FIXED JENKINS-1613&#93;</span> implements touchstone build for matrix projects, which runs a selected number of configurations first and then all the rests if these succeed</p>
</pre>
</details>

</ul>
</details>

<p>Code changed in hudson<br/>
User: : huybrechts<br/>
Path:<br/>
 trunk/hudson/main/core/src/main/java/hudson/matrix/MatrixBuild.java<br/>
 trunk/hudson/main/core/src/main/java/hudson/matrix/MatrixProject.java<br/>
 trunk/hudson/main/core/src/main/java/hudson/model/Result.java<br/>
 trunk/hudson/main/core/src/main/resources/hudson/matrix/MatrixProject/configure-entries.jelly<br/>
 trunk/hudson/main/test/src/main/java/org/jvnet/hudson/test/FailureBuilder.java<br/>
 trunk/hudson/main/test/src/main/java/org/jvnet/hudson/test/UnstableBuilder.java<br/>
 trunk/hudson/main/test/src/main/resources/org/jvnet/hudson/test/UnstableBuilder/config.jelly<br/>
 trunk/hudson/main/test/src/test/java/hudson/matrix/MatrixProjectTest.java<br/>
 trunk/hudson/main/war/resources/help/matrix/touchstone.html<br/>
<a href="http://fisheye4.cenqua.com/changelog/hudson/?cs=20737" class="external-link" target="_blank" rel="nofollow noopener">http://fisheye4.cenqua.com/changelog/hudson/?cs=20737</a><br/>
Log:<br/>
<span class="error">&#91;FIXED JENKINS-1613&#93;</span> implements touchstone build for matrix projects, which runs a selected number of configurations first and then all the rests if these succeed</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>huybrechts:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=124444&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-124444">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=124444&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-124444">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>oops, closed the wrong issue</p>
</pre>
</details>

</ul>
</details>

<p>oops, closed the wrong issue</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/lars_kruse.png" title="lars_kruse's avatar" /> lars_kruse:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=147705&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147705">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=147705&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147705">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I've assigned this case to myself.</p>

<p>We ('Christian Wolfgang' and myself) are researching and we will present a design proposal for a simple and elegant solution to this problem within a few days.</p>

<p>Hope to get feed-back on the design from the community before we start implementing.</p>

<p>Stay tuned!</p>
</pre>
</details>

</ul>
</details>

<p>I've assigned this case to myself.</p>

<p>We ('Christian Wolfgang' and myself) are researching and we will present a design proposal for a simple and elegant solution to this problem within a few days.</p>

<p>Hope to get feed-back on the design from the community before we start implementing.</p>

<p>Stay tuned!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/wolfgang.jpg" title="wolfgang's avatar" /> wolfgang:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=147824&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147824">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=147824&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147824">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Modifications to Jenkins-core
-----------------------------</p>

<p>The general idea of the design is being able to differentiate between matrix configurations reusing(copying) older builds(a basis build) and configurations actual being rebuild, without disturbing older non-reusing projects' configurations.
For this to work we need to alter the Jenkins core at the very lowest level, where the job is run. This is done in the abstract class Run, in the method run(). This method has two points of interest:</p>
<ul>
	<li>job.run(listener)</li>
	<li>job.post(listener)
Cutting of these two calls, we can achieve a build, that looks like it was run correctly, setting the build result as we want(=the same result as the basis build). Being able to create a Run object, whose values are set, we only need to copy the artifacts from the basis builds, the ones that are not rebuild. This can be achieved by adding some extra functionality to the AbstractBuild class. Currently, as it can be viewed in the code, the build directory of the basis build is hard copied to the source build, only leaving out build.xml. The build.xml file is created by Jenkins after the job is run and would cause an exception.</li>
</ul>


<p>To distinguish between builds that should be rebuild and builds that should not, we need to add a property to the Run class. We have added a small class called ReuseRun, that only implements an integer referencing the rebuild build. If this object is either null or the integer is 0, the new functionality is left untouched. This object is null for all projects not using the rebuild functionality.</p>

<p>Though this is a prototype, we consider this a functional starting point, based on the fact that Run objects will, by default, instantiate the ReuseRun object to null. Thus only explicit calls to setReuse can change that and the code added is only related to the object being not null.</p>
</pre>
</details>

</ul>
</details>

<p>Modifications to Jenkins-core<br/>
-----------------------------</p>

<p>The general idea of the design is being able to differentiate between matrix configurations reusing(copying) older builds(a basis build) and configurations actual being rebuild, without disturbing older non-reusing projects' configurations.<br/>
For this to work we need to alter the Jenkins core at the very lowest level, where the job is run. This is done in the abstract class Run, in the method run(). This method has two points of interest:</p>
<ul>
	<li>job.run(listener)</li>
	<li>job.post(listener)<br/>
Cutting of these two calls, we can achieve a build, that looks like it was run correctly, setting the build result as we want(=the same result as the basis build). Being able to create a Run object, whose values are set, we only need to copy the artifacts from the basis builds, the ones that are not rebuild. This can be achieved by adding some extra functionality to the AbstractBuild class. Currently, as it can be viewed in the code, the build directory of the basis build is hard copied to the source build, only leaving out build.xml. The build.xml file is created by Jenkins after the job is run and would cause an exception.</li>
</ul>


<p>To distinguish between builds that should be rebuild and builds that should not, we need to add a property to the Run class. We have added a small class called ReuseRun, that only implements an integer referencing the rebuild build. If this object is either null or the integer is 0, the new functionality is left untouched. This object is null for all projects not using the rebuild functionality.</p>

<p>Though this is a prototype, we consider this a functional starting point, based on the fact that Run objects will, by default, instantiate the ReuseRun object to null. Thus only explicit calls to setReuse can change that and the code added is only related to the object being not null.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/wolfgang.jpg" title="wolfgang's avatar" /> wolfgang:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=147825&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147825">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=147825&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147825">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I have committed the initial code draft at <a href="https://github.com/Praqma/jenkins/compare/master" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/Praqma/jenkins/compare/master</a></p>
</pre>
</details>

</ul>
</details>

<p>I have committed the initial code draft at <a href="https://github.com/Praqma/jenkins/compare/master" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/Praqma/jenkins/compare/master</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/wolfgang.jpg" title="wolfgang's avatar" /> wolfgang:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=147826&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147826">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=147826&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147826">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Btw the modifications to MatrixBuild is just empty lines and MatrixRun is just for my small mock up test.</p>
</pre>
</details>

</ul>
</details>

<p>Btw the modifications to MatrixBuild is just empty lines and MatrixRun is just for my small mock up test.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>dty:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=147841&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147841">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=147841&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147841">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>In the situation where the build you are re-using does not exist, you are failing the matrix build. Why is failing the build better than rebuilding that configuration?</p>

<p>I also see that you haven't provided UI changes yet. How do you think this functionality would be exposed? I imagine you could add a Rebuild action to the left side navigation, which would take the user to a page where they can pick the configuration subset they want to rebuild. Maybe there's another way to do this that I didn't consider.</p>

<p>A fundamental problem I see with this capability is that it requires a human to know that it's safe to only rebuild parts of the matrix. Going back to the original issue description, if the person doing the rebuild doesn't realize that the code change also affects another OS that doesn't get rebuilt, then you have failed to test a necessary configuration. This also then bubbles up into release management. You probably want to tag your version control at the point of release, but if some of your configurations were built with older versions and not rebuilt, you've lost that information.</p>

</pre>
</details>

</ul>
</details>

<p>In the situation where the build you are re-using does not exist, you are failing the matrix build. Why is failing the build better than rebuilding that configuration?</p>

<p>I also see that you haven't provided UI changes yet. How do you think this functionality would be exposed? I imagine you could add a Rebuild action to the left side navigation, which would take the user to a page where they can pick the configuration subset they want to rebuild. Maybe there's another way to do this that I didn't consider.</p>

<p>A fundamental problem I see with this capability is that it requires a human to know that it's safe to only rebuild parts of the matrix. Going back to the original issue description, if the person doing the rebuild doesn't realize that the code change also affects another OS that doesn't get rebuilt, then you have failed to test a necessary configuration. This also then bubbles up into release management. You probably want to tag your version control at the point of release, but if some of your configurations were built with older versions and not rebuilt, you've lost that information.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/lars_kruse.png" title="lars_kruse's avatar" /> lars_kruse:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=147843&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147843">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=147843&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-147843">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi Dean, thanks for feed back - Good questions.</p>

<p>Why fail rather than rebuild if the configuration is missing? Well we think like this: In the case where we need to make the choice, the user obviously asked for a reuse - not a rebuild. Potentially a rebuild can take a long time - so it's likely that our guess will not always capture the users intend or wish. Therefore we choose to fail the build and print to the log what happened - simply because it's faster, and this way we're not assuming anything. Then the user (yes - there IS a human involved in this feature) can schedule another rebuild, only this time he explicitly asks for a rebuild - not a reuse.</p>

<p>You are right about your assumptions regarding the UI (we will present a mock-up of what we have in mind very soon) but in large: A rebuild action that takes you to a new form listing all the results in the matrix from the previous build, each cell in the matrix now has a small checkbox attached to it. The user checks the builds in the matrix he/she wants to rebuild - the other ones are reused. There will be short cuts like "select all failed", "select all unstable" ...etc. When the user hits submit, the choices are traversed and all the builds that shall be reused will have the ReuseRun attribute set.</p>

<p>Our plan is to extend the <b>ability</b> to reuse a build in the core (as presented here) and then we will release the form that utilizes it as a seperate (new) plugin.</p>

<p>...leading to your third question - or concern: It requires a knowning human! Yes! The scenario we see for this reuse/rebuild feature is:
In a large matrix some (most) configurations has succeeded, a few have failed or are unstable. Going through the logs it appears that it's due to the slaves configuration or abilities that something went wrong (say a process hung or a network connection was lost or it failed to get a license to tool it depended on....) Now you don't want to rebuild the entire matrix again (imagine that it takes several hours to do that). You just pick the failed builds and ask to reuse the successful ones. We imagine that the configurations that are rerun will still honour all the various other settings of the job, thus if the job picks latest commit on a branch, and there is a new commit - then you are right it's a different build then before. </p>

<p>We imagine that users must take the necessary steps in their job configurations to avoid mistakes like this to happen. In our setup we're building a named commit, so it's safe - for us <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>The thing is, we realize that the behaviour we're looking for might be special, but the aproach to how we implemented it (in AbstractBuild and Run) is generic. Other plugins could use the same feature to make different behaviour - either by contributing to our plugin or by creating a new one. The design should support that no more extensions are needed in the core to make either work.</p>

<p>Hmmm - But to avoid making future changes to the core on behalf of others in the future would rather rebuild than fail the build, then I guess we could extend the ReuseRun object to be able to carry such information (rebuildIfMissing) and act accordingly. We could set it to FALSE as default and get the same behaviour. Would that work?    </p>

<p>Again, Thanks for feed back, it's very welcome!</p>
</pre>
</details>

</ul>
</details>

<p>Hi Dean, thanks for feed back - Good questions.</p>

<p>Why fail rather than rebuild if the configuration is missing? Well we think like this: In the case where we need to make the choice, the user obviously asked for a reuse - not a rebuild. Potentially a rebuild can take a long time - so it's likely that our guess will not always capture the users intend or wish. Therefore we choose to fail the build and print to the log what happened - simply because it's faster, and this way we're not assuming anything. Then the user (yes - there IS a human involved in this feature) can schedule another rebuild, only this time he explicitly asks for a rebuild - not a reuse.</p>

<p>You are right about your assumptions regarding the UI (we will present a mock-up of what we have in mind very soon) but in large: A rebuild action that takes you to a new form listing all the results in the matrix from the previous build, each cell in the matrix now has a small checkbox attached to it. The user checks the builds in the matrix he/she wants to rebuild - the other ones are reused. There will be short cuts like "select all failed", "select all unstable" ...etc. When the user hits submit, the choices are traversed and all the builds that shall be reused will have the ReuseRun attribute set.</p>

<p>Our plan is to extend the <b>ability</b> to reuse a build in the core (as presented here) and then we will release the form that utilizes it as a seperate (new) plugin.</p>

<p>...leading to your third question - or concern: It requires a knowning human! Yes! The scenario we see for this reuse/rebuild feature is:<br/>
In a large matrix some (most) configurations has succeeded, a few have failed or are unstable. Going through the logs it appears that it's due to the slaves configuration or abilities that something went wrong (say a process hung or a network connection was lost or it failed to get a license to tool it depended on....) Now you don't want to rebuild the entire matrix again (imagine that it takes several hours to do that). You just pick the failed builds and ask to reuse the successful ones. We imagine that the configurations that are rerun will still honour all the various other settings of the job, thus if the job picks latest commit on a branch, and there is a new commit - then you are right it's a different build then before. </p>

<p>We imagine that users must take the necessary steps in their job configurations to avoid mistakes like this to happen. In our setup we're building a named commit, so it's safe - for us <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p>The thing is, we realize that the behaviour we're looking for might be special, but the aproach to how we implemented it (in AbstractBuild and Run) is generic. Other plugins could use the same feature to make different behaviour - either by contributing to our plugin or by creating a new one. The design should support that no more extensions are needed in the core to make either work.</p>

<p>Hmmm - But to avoid making future changes to the core on behalf of others in the future would rather rebuild than fail the build, then I guess we could extend the ReuseRun object to be able to carry such information (rebuildIfMissing) and act accordingly. We could set it to FALSE as default and get the same behaviour. Would that work?    </p>

<p>Again, Thanks for feed back, it's very welcome!</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=148668&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148668">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=148668&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148668">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Christian Wolfgang
Path:
 core/src/main/java/hudson/matrix/MatrixBuild.java
 core/src/main/java/hudson/matrix/listeners/MatrixBuildListener.java
<a href="http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4</a>
Log:
More on <a href="https://issue-redirect.jenkins.io/issue/1613" title="Ability to build on configurable subset of matrix configurations" class="issue-link" data-issue-key="JENKINS-1613"><del>JENKINS-1613</del></a></p>



</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Christian Wolfgang<br/>
Path:<br/>
 core/src/main/java/hudson/matrix/MatrixBuild.java<br/>
 core/src/main/java/hudson/matrix/listeners/MatrixBuildListener.java<br/>
<a href="http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4</a><br/>
Log:<br/>
More on <a href="https://issue-redirect.jenkins.io/issue/1613" title="Ability to build on configurable subset of matrix configurations" class="issue-link" data-issue-key="JENKINS-1613"><del>JENKINS-1613</del></a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=148684&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148684">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=148684&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148684">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Kohsuke Kawaguchi
Path:
 changelog.html
<a href="http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48</a>
Log:
<span class="error">&#91;FIXED JENKINS-1613&#93;</span> Ability to build on configurable subset of matrix
configurations</p>

<p>This change introduces the necessary extension point. The implementation
of this will be in a plugin.</p>


<p>Compare: <a href="https://github.com/jenkinsci/jenkins/compare/f10b5eb...0203876" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/compare/f10b5eb...0203876</a></p>

</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Kohsuke Kawaguchi<br/>
Path:<br/>
 changelog.html<br/>
<a href="http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48</a><br/>
Log:<br/>
<span class="error">&#91;FIXED JENKINS-1613&#93;</span> Ability to build on configurable subset of matrix<br/>
configurations</p>

<p>This change introduces the necessary extension point. The implementation<br/>
of this will be in a plugin.</p>


<p>Compare: <a href="https://github.com/jenkinsci/jenkins/compare/f10b5eb...0203876" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/compare/f10b5eb...0203876</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=148685&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148685">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=148685&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148685">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Christian Wolfgang
Path:
 core/src/main/java/hudson/matrix/MatrixBuild.java
 core/src/main/java/hudson/matrix/listeners/MatrixBuildListener.java
<a href="http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4</a>
Log:
More on <a href="https://issue-redirect.jenkins.io/issue/1613" title="Ability to build on configurable subset of matrix configurations" class="issue-link" data-issue-key="JENKINS-1613"><del>JENKINS-1613</del></a></p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Christian Wolfgang<br/>
Path:<br/>
 core/src/main/java/hudson/matrix/MatrixBuild.java<br/>
 core/src/main/java/hudson/matrix/listeners/MatrixBuildListener.java<br/>
<a href="http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4</a><br/>
Log:<br/>
More on <a href="https://issue-redirect.jenkins.io/issue/1613" title="Ability to build on configurable subset of matrix configurations" class="issue-link" data-issue-key="JENKINS-1613"><del>JENKINS-1613</del></a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=148686&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148686">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=148686&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148686">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Kohsuke Kawaguchi
Path:
 changelog.html
<a href="http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48</a>
Log:
<span class="error">&#91;FIXED JENKINS-1613&#93;</span> Ability to build on configurable subset of matrix
configurations</p>

<p>This change introduces the necessary extension point. The implementation
of this will be in a plugin.</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Kohsuke Kawaguchi<br/>
Path:<br/>
 changelog.html<br/>
<a href="http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48</a><br/>
Log:<br/>
<span class="error">&#91;FIXED JENKINS-1613&#93;</span> Ability to build on configurable subset of matrix<br/>
configurations</p>

<p>This change introduces the necessary extension point. The implementation<br/>
of this will be in a plugin.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/dogfood.png" title="dogfood's avatar" /> dogfood:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=148691&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148691">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=148691&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148691">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/yellow.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/791/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #791</a>
     More on <a href="https://issue-redirect.jenkins.io/issue/1613" title="Ability to build on configurable subset of matrix configurations" class="issue-link" data-issue-key="JENKINS-1613"><del>JENKINS-1613</del></a>
<span class="error">&#91;FIXED JENKINS-1613&#93;</span> Ability to build on configurable subset of matrix</p>

<p>Christian Wolfgang : <a href="https://github.com/jenkinsci/jenkins/commit/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4" class="external-link" target="_blank" rel="nofollow noopener">f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4</a>
Files : </p>
<ul>
	<li>core/src/main/java/hudson/matrix/listeners/MatrixBuildListener.java</li>
	<li>core/src/main/java/hudson/matrix/MatrixBuild.java</li>
</ul>


<p>Kohsuke Kawaguchi : <a href="https://github.com/jenkinsci/jenkins/commit/0203876cb01f0cfee1c790a3102a3508ff4f6a48" class="external-link" target="_blank" rel="nofollow noopener">0203876cb01f0cfee1c790a3102a3508ff4f6a48</a>
Files : </p>
<ul>
	<li>changelog.html</li>
</ul>

</pre>
</details>

</ul>
</details>

<p>Integrated in <span class="image-wrap" style=""><img src="http://ci.jenkins-ci.org/images/16x16/yellow.png" style="border: 0px solid black" /></span> <a href="http://ci.jenkins-ci.org/job/jenkins_main_trunk/791/" class="external-link" target="_blank" rel="nofollow noopener">jenkins_main_trunk #791</a><br/>
     More on <a href="https://issue-redirect.jenkins.io/issue/1613" title="Ability to build on configurable subset of matrix configurations" class="issue-link" data-issue-key="JENKINS-1613"><del>JENKINS-1613</del></a><br/>
<span class="error">&#91;FIXED JENKINS-1613&#93;</span> Ability to build on configurable subset of matrix</p>

<p>Christian Wolfgang : <a href="https://github.com/jenkinsci/jenkins/commit/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4" class="external-link" target="_blank" rel="nofollow noopener">f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4</a><br/>
Files : </p>
<ul>
	<li>core/src/main/java/hudson/matrix/listeners/MatrixBuildListener.java</li>
	<li>core/src/main/java/hudson/matrix/MatrixBuild.java</li>
</ul>


<p>Kohsuke Kawaguchi : <a href="https://github.com/jenkinsci/jenkins/commit/0203876cb01f0cfee1c790a3102a3508ff4f6a48" class="external-link" target="_blank" rel="nofollow noopener">0203876cb01f0cfee1c790a3102a3508ff4f6a48</a><br/>
Files : </p>
<ul>
	<li>changelog.html</li>
</ul>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/wolfgang.jpg" title="wolfgang's avatar" /> wolfgang:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=148988&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148988">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=148988&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148988">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Given the problems arisen by rebuilt matrices composed of several different matrix builds, turns out to be too big.
Thus the demand for rebuilding a matrix from a specific matrix build is required.
This requirement needs to be build into the core.</p>
</pre>
</details>

</ul>
</details>

<p>Given the problems arisen by rebuilt matrices composed of several different matrix builds, turns out to be too big.<br/>
Thus the demand for rebuilding a matrix from a specific matrix build is required.<br/>
This requirement needs to be build into the core.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/wolfgang.jpg" title="wolfgang's avatar" /> wolfgang:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=148989&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148989">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=148989&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-148989">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I have submitted a pull request fixing this issue: <a href="https://github.com/jenkinsci/jenkins/pull/133" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/133</a></p>
</pre>
</details>

</ul>
</details>

<p>I have submitted a pull request fixing this issue: <a href="https://github.com/jenkinsci/jenkins/pull/133" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/jenkinsci/jenkins/pull/133</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>brian:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=149414&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149414">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=149414&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149414">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote>
<p>thus if the job picks latest commit on a branch, and there is a new commit - then you are right it's a different build then before.</p></blockquote>

<p>Doesn't every job (that builds from an SCM) have some sort of "bookmark" (i.e. the commit hash in git) of the point in the SCM's history where a build was done?  Couldn't this "rebuild" feature ensure that the SCM checkout is done at that "bookmark" rather than the <em>latest</em> as it usually is supposed to do?</p>

<p>That would then ensure that you don't have a different build than the one you are rebuilding.</p>
</pre>
</details>

</ul>
</details>

<blockquote>
<p>thus if the job picks latest commit on a branch, and there is a new commit - then you are right it's a different build then before.</p></blockquote>

<p>Doesn't every job (that builds from an SCM) have some sort of "bookmark" (i.e. the commit hash in git) of the point in the SCM's history where a build was done?  Couldn't this "rebuild" feature ensure that the SCM checkout is done at that "bookmark" rather than the <em>latest</em> as it usually is supposed to do?</p>

<p>That would then ensure that you don't have a different build than the one you are rebuilding.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>brian:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=149415&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149415">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=149415&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149415">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is there any projected time frame in which UI exposing this feature will be available?</p>
</pre>
</details>

</ul>
</details>

<p>Is there any projected time frame in which UI exposing this feature will be available?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/wolfgang.jpg" title="wolfgang's avatar" /> wolfgang:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=149416&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149416">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=149416&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149416">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Yes, the core changes are in 1.416, so my guess is this upcoming Monday.
I will release the plugin when 1.416 is released.</p>
</pre>
</details>

</ul>
</details>

<p>Yes, the core changes are in 1.416, so my guess is this upcoming Monday.<br/>
I will release the plugin when 1.416 is released.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>brian:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=149422&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149422">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=149422&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149422">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<blockquote>
<p>Yes, the core changes are in 1.416, so my guess is this upcoming Monday.</p></blockquote>

<p>Oh.  I thought this went into 1.413.  From the Changelog:</p>

<p>What's new in 1.413 (2011/05/22)
...
Support rebuilding a subset of matrix configurations (issue 1613)
...</p>

<p>Was the commit to 1.413 incomplete perhaps?</p>
</pre>
</details>

</ul>
</details>

<blockquote>
<p>Yes, the core changes are in 1.416, so my guess is this upcoming Monday.</p></blockquote>

<p>Oh.  I thought this went into 1.413.  From the Changelog:</p>

<p>What's new in 1.413 (2011/05/22)<br/>
...<br/>
Support rebuilding a subset of matrix configurations (issue 1613)<br/>
...</p>

<p>Was the commit to 1.413 incomplete perhaps?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/wolfgang.jpg" title="wolfgang's avatar" /> wolfgang:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=149428&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149428">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=149428&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-149428">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Somewhat. The 1.413 always bases its rebuild from the latest successful build, the new change can base from any build.</p>
</pre>
</details>

</ul>
</details>

<p>Somewhat. The 1.413 always bases its rebuild from the latest successful build, the new change can base from any build.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=152715&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-152715">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=152715&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-152715">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Christian Wolfgang
Path:
 core/src/main/java/hudson/matrix/MatrixBuild.java
 core/src/main/java/hudson/matrix/listeners/MatrixBuildListener.java
<a href="http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4</a>
Log:
  More on <a href="https://issue-redirect.jenkins.io/issue/1613" title="Ability to build on configurable subset of matrix configurations" class="issue-link" data-issue-key="JENKINS-1613"><del>JENKINS-1613</del></a></p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Christian Wolfgang<br/>
Path:<br/>
 core/src/main/java/hudson/matrix/MatrixBuild.java<br/>
 core/src/main/java/hudson/matrix/listeners/MatrixBuildListener.java<br/>
<a href="http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4</a><br/>
Log:<br/>
  More on <a href="https://issue-redirect.jenkins.io/issue/1613" title="Ability to build on configurable subset of matrix configurations" class="issue-link" data-issue-key="JENKINS-1613"><del>JENKINS-1613</del></a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=152716&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-152716">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=152716&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-152716">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Kohsuke Kawaguchi
Path:
 changelog.html
<a href="http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48</a>
Log:
  <span class="error">&#91;FIXED JENKINS-1613&#93;</span> Ability to build on configurable subset of matrix
configurations</p>

<p>This change introduces the necessary extension point. The implementation
of this will be in a plugin.</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Kohsuke Kawaguchi<br/>
Path:<br/>
 changelog.html<br/>
<a href="http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-1613&#93;</span> Ability to build on configurable subset of matrix<br/>
configurations</p>

<p>This change introduces the necessary extension point. The implementation<br/>
of this will be in a plugin.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=154230&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-154230">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=154230&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-154230">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Christian Wolfgang
Path:
 core/src/main/java/hudson/matrix/MatrixBuild.java
 core/src/main/java/hudson/matrix/listeners/MatrixBuildListener.java
<a href="http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4</a>
Log:
  More on <a href="https://issue-redirect.jenkins.io/issue/1613" title="Ability to build on configurable subset of matrix configurations" class="issue-link" data-issue-key="JENKINS-1613"><del>JENKINS-1613</del></a></p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Christian Wolfgang<br/>
Path:<br/>
 core/src/main/java/hudson/matrix/MatrixBuild.java<br/>
 core/src/main/java/hudson/matrix/listeners/MatrixBuildListener.java<br/>
<a href="http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4</a><br/>
Log:<br/>
  More on <a href="https://issue-redirect.jenkins.io/issue/1613" title="Ability to build on configurable subset of matrix configurations" class="issue-link" data-issue-key="JENKINS-1613"><del>JENKINS-1613</del></a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=154231&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-154231">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=154231&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-154231">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Kohsuke Kawaguchi
Path:
 changelog.html
<a href="http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48</a>
Log:
  <span class="error">&#91;FIXED JENKINS-1613&#93;</span> Ability to build on configurable subset of matrix
configurations</p>

<p>This change introduces the necessary extension point. The implementation
of this will be in a plugin.</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Kohsuke Kawaguchi<br/>
Path:<br/>
 changelog.html<br/>
<a href="http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/jenkins/0203876cb01f0cfee1c790a3102a3508ff4f6a48</a><br/>
Log:<br/>
  <span class="error">&#91;FIXED JENKINS-1613&#93;</span> Ability to build on configurable subset of matrix<br/>
configurations</p>

<p>This change introduces the necessary extension point. The implementation<br/>
of this will be in a plugin.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>salimfadhley:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=155828&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-155828">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=155828&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-155828">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>One problem with the current implementation is it's effect on the build history graph - suppose you incrementally re-run parts of an unreliable matrix build, it gives the impression that the number of tests in each run are diminishing - when in actual fact the number of tests remain constant, we are just running a smaller subset.</p>

<p>Would it be possible to alter this so that results from previous runs which are not re-executed by Matrix Reloaded are carried forwards. That means I'd like to see the height of the bar-graph remain constant with the red proportion gradually diminishing with each additional build. </p>

<p>Here's a slightly exaggerated screenshot which illustrates the problem:</p>

<p><a href="http://imgur.com/HEV9J" class="external-link" target="_blank" rel="nofollow noopener">http://imgur.com/HEV9J</a></p>
</pre>
</details>

</ul>
</details>

<p>One problem with the current implementation is it's effect on the build history graph - suppose you incrementally re-run parts of an unreliable matrix build, it gives the impression that the number of tests in each run are diminishing - when in actual fact the number of tests remain constant, we are just running a smaller subset.</p>

<p>Would it be possible to alter this so that results from previous runs which are not re-executed by Matrix Reloaded are carried forwards. That means I'd like to see the height of the bar-graph remain constant with the red proportion gradually diminishing with each additional build. </p>

<p>Here's a slightly exaggerated screenshot which illustrates the problem:</p>

<p><a href="http://imgur.com/HEV9J" class="external-link" target="_blank" rel="nofollow noopener">http://imgur.com/HEV9J</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=163294&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-163294">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=163294&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-163294">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: Christian Wolfgang
Path:
 core/src/main/java/hudson/matrix/MatrixBuild.java
 core/src/main/java/hudson/matrix/listeners/MatrixBuildListener.java
<a href="http://jenkins-ci.org/commit/matrix-project-plugin/1b135d4cbc08da6959faf63842023962276675c8" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/matrix-project-plugin/1b135d4cbc08da6959faf63842023962276675c8</a>
Log:
  More on <a href="https://issue-redirect.jenkins.io/issue/1613" title="Ability to build on configurable subset of matrix configurations" class="issue-link" data-issue-key="JENKINS-1613"><del>JENKINS-1613</del></a></p>

<p>Originally-Committed-As: f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: Christian Wolfgang<br/>
Path:<br/>
 core/src/main/java/hudson/matrix/MatrixBuild.java<br/>
 core/src/main/java/hudson/matrix/listeners/MatrixBuildListener.java<br/>
<a href="http://jenkins-ci.org/commit/matrix-project-plugin/1b135d4cbc08da6959faf63842023962276675c8" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/matrix-project-plugin/1b135d4cbc08da6959faf63842023962276675c8</a><br/>
Log:<br/>
  More on <a href="https://issue-redirect.jenkins.io/issue/1613" title="Ability to build on configurable subset of matrix configurations" class="issue-link" data-issue-key="JENKINS-1613"><del>JENKINS-1613</del></a></p>

<p>Originally-Committed-As: f10b5eb4cd2d1d1fd68f0c469f7a9a78e45c68a4</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>rayeesn:</i></summary>

<ul>
<li><i><a class="no-jira-link-rewrite" href="https://issue-redirect.jenkins.io/issue/1613?focusedId=180268&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-180268">Comment link</a></i>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-1613?focusedId=180268&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-180268">comment link</a> (no redirect)</i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Subset start jobs with new Run number,  it will generate separate report;</p>

<p>Is it possible to make option to append result with its parent job number ?</p>

<p>Example Executed Job 100, there are 5 failure, if you rebuild the subset, starting job from 101; instead of this can append this with 100 itself ? </p>
</pre>
</details>

</ul>
</details>

<p>Subset start jobs with new Run number,  it will generate separate report;</p>

<p>Is it possible to make option to append result with its parent job number ?</p>

<p>Example Executed Job 100, there are 5 failure, if you rebuild the subset, starting job from 101; instead of this can append this with 100 itself ? </p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/matrix-project-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-2353%22">JENKINS-2353</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `is_duplicated_by` from Jira: <a href="https://github.com/jenkinsci/matrix-project-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-3867%22">JENKINS-3867</a>]</i>
